phoenix-core/src/main/java/org/apache/phoenix/schema/transform/Transform.java

```diff
@@ -157,64 +307,70 @@ public static void addTransform(
+        String indexTableName = SchemaUtil.getTableNameFromFullName(PhoenixConfigurationUtil.getIndexToolIndexTableName(configuration));
+        String logicaTableName = oldTableLogicalName;
+        String logicalParentName = null;
+        if (PhoenixConfigurationUtil.getTransformingTableType(configuration) == IndexScrutinyTool.SourceTable.INDEX_TABLE_SOURCE)
+            if (!Strings.isNullOrEmpty(indexTableName)) {
+                logicaTableName = indexTableName;
+                logicalParentName = SchemaUtil.getTableName(schemaName, oldTableLogicalName);
             }
-            return null;
-        }
-    }
-
-    public static boolean checkIsTransformNeeded(MetaDataClient.MetaProperties metaProperties, String schemaName,
-                                                 PTable table, String logicalTableName, String parentTableName,
-                                                 String tenantId, PhoenixConnection connection) throws SQLException {
-        boolean isTransformNeeded = isTransformNeeded(metaProperties, table);
-        if (isTransformNeeded) {
-            SystemTransformRecord existingTransform = Transform.getTransformRecord(schemaName, logicalTableName, parentTableName, tenantId,connection);
-            if (existingTransform != null && existingTransform.isActive()) {
-                throw new SQLExceptionInfo.Builder(
-                        SQLExceptionCode.CANNOT_TRANSFORM_ALREADY_TRANSFORMING_TABLE)
-                        .setMessage(" Only one transform at a time is allowed ")
-                        .setSchemaName(schemaName).setTableName(logicalTableName).build().buildException();
+        boolean isPartial = PhoenixConfigurationUtil.getIsPartialTransform(configuration);
+        SystemTransformRecord transformRecord = getTransformRecord(schemaName, logicaTableName, logicalParentName,
+                tenantId, connection.unwrap(PhoenixConnection.class));
+        if (!isPartial) {
+            String newTableName = SchemaUtil.getTableNameFromFullName(transformRecord.getNewPhysicalTableName());
+            PTable pNewTable = PhoenixRuntime.getTable(connection, transformRecord.getNewPhysicalTableName());
+            PTable pOldTable = PhoenixRuntime.getTable(connection, SchemaUtil.getTableName(schemaName,logicaTableName));
+            if (pOldTable.getImmutableStorageScheme() != pNewTable.getImmutableStorageScheme() ||
+                    pOldTable.getEncodingScheme() != pNewTable.getEncodingScheme()) {
+                MetaDataClient.mutateTransformProperties(connection, tenantId, schemaName, logicaTableName, newTableName,
+                        pNewTable.getImmutableStorageScheme(), pNewTable.getEncodingScheme());
+                // We need to update the columns's qualifiers as well
+                if (pOldTable.getEncodingScheme() != pNewTable.getEncodingScheme()) {
+                    Short nextKeySeq = 0;
```

Comment:
nit: primitive `short`?

Reply:
addColumnMutation checks for null value for keySeq.